src/main/java/service/GitLabService.java

```diff
@@ -504,31 +511,38 @@ private boolean haveDiff(String gitlabEventUUID, Long project, String from, Stri
 		}
 	}
 
-	private Long getCascadeResponsible(String gitlabEventUUID, Long project, Long mrNumber) {
+	private List<Long> getCascadeResponsibleIds(String gitlabEventUUID, Long project, Long mrNumber) {
 		Log.infof("GitlabEvent: '%s' | Backtracing cascade responsible: get merge request '!%d' in project '%d'", gitlabEventUUID, mrNumber, project);
 		MergeRequest mr = getMr(gitlabEventUUID, project, mrNumber);
-		if (mr != null) {
-			Long mrMerger = mr.getMergeUser().getId();
-			Log.infof("GitlabEvent: '%s' | MR merger has id '%d', ucascade user has id '%d'", gitlabEventUUID, mrMerger, ucascadeUser);
-
-			if (Objects.equals(mrMerger, ucascadeUser)) {
-				Long prevMRNumber = getPrevMergeRequestNumber(mr.getSourceBranch());
-				if (prevMRNumber == null) {
-					Log.warnf("GitlabEvent: '%s' | Cannot get cascade responsible of merge request '!%d' in project '%d' (could not compute the previous MR)", gitlabEventUUID, mrNumber, project);
-					return null;
-				} else {
-					Long cascadeResponsibleId = getCascadeResponsible(gitlabEventUUID, project, prevMRNumber);
-					if (cascadeResponsibleId == null) {
-						Log.warnf("GitlabEvent: '%s' | Cannot get cascade responsible of merge request '!%d' in project '%d'", gitlabEventUUID, mrNumber, project);
-					}
-					return cascadeResponsibleId;
-				}
-			} else {
-				return mrMerger;
+		if (mr == null) {
+			return null;
+		}
+
+		Long mrMerger = mr.getMergeUser().getId();
+		Log.infof("GitlabEvent: '%s' | MR merger has id '%d', ucascade user has id '%d'", gitlabEventUUID, mrMerger, ucascadeUser);
+
+		if (!Objects.equals(mrMerger, ucascadeUser)) {
+			// success exit point, found the origin mr
+			List<Long> cascadeResponsibleIds = mr.getAssignees().stream().map(Assignee::getId).collect(Collectors.toList());
+			if (!cascadeResponsibleIds.contains(mrMerger)) {
+				cascadeResponsibleIds.add(mrMerger);
```

Comment:
> I don't think that the list obtained with .collect(Collectors.toList()) is mutable.

I am using `.collect(Collectors.toList())` instead of just the `.toList()` (from Stream) precisely to make it mutable. Maybe you are confusing the two? 

I might be missing something but this seems to work well. Are you thinking of a special case?

Ref: https://www.javacodegeeks.com/2020/12/jdk-16-stream-to-list-in-one-easy-call.html

Reply:
Ok for me both were equivalent… I thought `Collectors.toList()` is creating an immutable list.